### ISSUE TITLE
Question about Sample 'Out of Range' Error 

### ISSUE THREAD

**Issue by PatrickRitchie** [mtconnect/cppagent#39]
Using the application sequence (recommended in section 3.2 of the MTConnect Components spec) of first running a Current request to get the FirstSequence or NextSequence parameter to then use in a subsequent Sample request can sometimes return an "OUT_OF_RANGE" error. 

I suspect this is due to the agent's buffer being full and the buffer setting is not set large enough but I was wondering if this should instead return the first available sequence instead of returning an error. This seems to be the way that the Count parameter works as it just returns the last available sequence when it exceeds the LastSequence value.

This typically isn't an issue when an application has already been running and it is only requesting a small range of the most recent data but can become an issue for applications that return the entire buffer. In this case, when the buffer is full, the application may end up in a situation where it is constantly chasing a moving target as by the time the application receives the Current request, the FirstSequence number is already invalid.

I haven't extensively looked at the source code for this agent but a quick observation points the issue (if this is indeed considered an issue) to the below code:

agent.cpp | Agent::handleCall()

```
if (start == NO_START) // If there was no data in queries
      {
        start = checkAndGetParam64(queries, "from", 1,
                                   getFirstSequence(), true, mSequence);
}
```

agent.cpp | Agent::checkAndGetParam() and Agent::checkAndGetParam64()

```
if (minValue != NO_VALUE64 && value < minValue)
  {
    if (minError)
    {
      throw ParameterError("OUT_OF_RANGE",
                          "'" + param + "' must be greater than or equal to " + intToString(minValue) + ".");
    }
    return minValue;
}
```

It appears the **minError** variable was meant to accomplish this but when used to process the Sample's start parameter it is set to **true**. Is there a reasoning behind this that it was set this way or is it something that can be updated? Or can this be a setting in the Agent's configuration file?

Thanks,

Patrick

**Comment by wsobel** [comment]
MTConnect uses a circular buffer and once if fills, the oldest data is removed. This means that if you request the FirstSequence number, it will most likely fail since it has already been removed as the data comes in. 

Let me consider the suggestion regarding defaulting to the fist if it is less than the first sequence. You make a good point regarding the behavior.

**Comment by wsobel** [comment]
I'm assuming this is closed.

**Comment by nicktavis** [comment]
I am still seeing this error. I've used the following python script to recreate it. I frequently get this output, indicating an error:

```
Next sequence: 1694397
'from' must be greater than or equal to 1694396
```

from lxml import etree

bufferSize = 333
base = 'http://<IP ADDRESS>:4076'
tree = etree.parse(f'{base}/sample?count=333').getroot()
header = tree.find( './/Header', namespaces=tree.nsmap )
agentNextSequence = header.attrib['nextSequence']

print(f'Next sequence from sample?count=333 {agentNextSequence}')

url = f'{base}/sample?from={agentNextSequence}&count={bufferSize}'
tree = etree.parse(url).getroot()
error = tree.findall( './/{' + tree.nsmap['m'] + '}Error' )
for err in error:
   print(err.text)

**Comment by wsobel** [comment]
I’ll check it out, it may be a one off error or a timing issue. Thanks, - W

**Comment by nicktavis** [comment]
Thank you for the quick response! We are testing against a 1.2 agent. The issue seems to be resolved for the 1.4 agent.

**Comment by wsobel** [comment]
Please try the latest version and see if there are issues. 1.2 is VERY old.
